glib: Bind `g_unichar` APIs

RealKC opened the pull request:
I'm not sure if documentation for these methods can be automatically generated given these are manually defined as methods on an extension trait.

I *think* I got all of the ones not offered by the stdlib

Closes #763 

---

Some notes:
* names were made "rustier", e.g.:` g_unichar_istitle` became `is_titlecase`
* I did not include APIs that were already availabe through Rust's char

RealKC commented on glib/src/unichar.rs:
```diff
@@ -0,0 +1,237 @@
+
+// rustdoc-stripper-ignore-next
+/// The result of a single step of the Unicode canonical decomposition algorithm
+pub enum DecompositionResult {
```
I'm not too happy with the names, but I don't have better ideas

sdroege replied:
Maybe `CharacterDecomposition`?

RealKC replied:
Yeah I liked that name more

RealKC commented on glib/src/unichar.rs:
```diff
@@ -0,0 +1,237 @@
+    fn fully_decompose(self, decomposition_kind: DecompositionKind) -> Vec<char> {
+        let buffer_len = ffi::G_UNICHAR_MAX_DECOMPOSITION_LENGTH as usize;
+        let mut buffer = vec![char::from(0); buffer_len];
+        let decomposition_length = unsafe {
+            ffi::g_unichar_fully_decompose(
+                self as _,
+                // NOTE: This assumes that glib will only ever write valid unicode codepoints to the
+                //       buffer.
+                decomposition_kind.into_bool().into_glib(),
+                buffer.as_mut_ptr().cast(),
+                buffer_len,
+            )
+        };
+        debug_assert!(decomposition_length <= buffer_len);
+        unsafe { buffer.set_len(decomposition_length) };
```
This feels kinda sketch but I think it's fine?

sdroege replied:
Yes but I would go one step further. Use `Vec::with_capacity()` above to create the `Vec` (and don't zero-initialize it).

bilelmoussaoui commented on glib/src/auto/enums.rs:
```diff
@@ -1052,6 +1052,261 @@ impl FromGlib<ffi::GTimeType> for TimeType {
+    #[doc(alias = "G_UNICODE_BREAK_HANGUL_LVT_SYLLABLE")]
+    HangulLvtSyllable,
+    #[doc(alias = "G_UNICODE_BREAK_CLOSE_PARANTHESIS")]
+    CloseParanthesis,
```
You will have to manually implement the enum, sadly. Unless you go and fix gir :)

RealKC replied:
I think I'll manually implement the enum then ^^

RealKC replied:
Oh that worked, thanks

bilelmoussaoui commented on glib/src/unichar.rs:
```diff
@@ -0,0 +1,239 @@
+    Canonical,
+}
+
+impl DecompositionKind {
+    fn into_bool(self) -> bool {
+        use DecompositionKind::*;
+        match self {
+            Compatibility => true,
+            Canonical => false,
+        }
+    }
+}
```
Why not use the From trait here?

RealKC replied:
I can't really recall my thought process here, I'll go with whatever you prefer

sdroege replied:
What is the use case for this?

RealKC replied:
I ended up just removing and inlining the contents of this function in the one place where it is used (`fully_decompose`)

bilelmoussaoui commented on glib/src/unichar.rs:
```diff
@@ -0,0 +1,239 @@
+    }
+
+    fn mirror_char(self) -> Option<char> {
+        let mut mirrored = 0u32;
```
We usually use mem::MaybeUninit here

RealKC replied:
I don't think it'd make that much of a difference?

sdroege replied:
By itself not: just avoids one instruction here (and in every inlined copy). But if everything did this, there would be measurable code bloat.

bilelmoussaoui commented on glib/src/unichar.rs:
```diff
@@ -0,0 +1,239 @@
+    }
+
+    fn is_mark(self) -> bool {
+        unsafe { bool::from_glib(ffi::g_unichar_ismark(self as _)) }
```
```suggestion
        unsafe { from_glib(ffi::g_unichar_ismark(self as _)) }
```
should be good enough, same for others

bilelmoussaoui commented on glib/src/unichar.rs:
```diff
@@ -0,0 +1,239 @@
+/// provided by GLib that do not exist in the standard library
+#[doc(alias = "g_unichar")]
+pub trait Unichar: sealed::Sealed + Copy {
+    #[doc(alias = "g_unichar_type")]
```
All these functions should be marked as manual in the Gir.toml

RealKC replied:
Will do

sdroege commented on glib/src/unichar.rs:
```diff
@@ -0,0 +1,239 @@
+
+// rustdoc-stripper-ignore-next
+/// The kind of decomposition to perform
+#[derive(Clone, Copy)]
```
Why not `PartialEq`, `Eq` and `Debug`? Those are useful :)

sdroege commented on glib/src/unichar.rs:
```diff
@@ -0,0 +1,239 @@
+
+// rustdoc-stripper-ignore-next
+/// The result of a single step of the Unicode canonical decomposition algorithm
+pub enum DecompositionResult {
```
Please derive useful traits here

sdroege commented on glib/src/unichar.rs:
```diff
@@ -0,0 +1,239 @@
+    #[doc(alias = "g_unichar_combining_class")]
+    #[doc(alias = "unichar_combining_class")]
+    // There's no enum defining named constants for this one as far as I can tell
+    fn combining_class(self) -> i32;
```
You could create one :)

RealKC replied:
There's quite a few of these ([UAX #44](https://www.unicode.org/reports/tr44/)], section 5.7.4 talks about them, [PropertyValueAliases.txt](https://www.unicode.org/Public/15.0.0/ucd/PropertyValueAliases.txt) (search for `# Canonical_Combining_Class (ccc)`) has a more complete list). Should we have an enum that has separate variants for the named ones, with the unnamed ones going into a generic `Ccc` variant? The gtk demos seem to assign more meaning to some combining classes but I'm not sure where they get that from. (see https://github.com/GNOME/gtk/blob/fa4ad1590dbbcaf1a0eb230d2a20d0a867f8b752/demos/gtk-demo/unicode-names.c#L101)

sdroege replied:
Not sure we should do that. We would have to keep that enum up to date :) I wonder why this is not already done in GLib though...

Let's keep it a plain integer then for now (but maybe it should be unsigned?)

RealKC replied:
I think the best type for this might be u8?

sdroege replied:
If it's only a byte and can't be more than that, sure

sdroege commented on glib/src/unichar.rs:
```diff
@@ -0,0 +1,239 @@
+fn from_unichar(unichar: u32) -> char {
+    // It's a glib bug if they give us an invalid unicode codepoint from one of their manipulation
+    // functions. It's safest if we just panic in that case
+    char::from_u32(unichar).unwrap()
```
That's just what `char::from_glib()` is already doing or not?

RealKC replied:
That doesn't seem to exist

bilelmoussaoui replied:
there is  char::try_from_glib though

RealKC replied:
I don't think so? I can't find an impl anyway: https://github.com/search?q=repo%3Agtk-rs%2Fgtk-rs-core%20%22impl%20TryFromGlib%22&type=code

bilelmoussaoui replied:
Oh it seems to use TryFrom instead, see https://github.com/gtk-rs/gtk4-rs/blob/db7af24294c706136f3bcba74caf284957b9b9c2/gtk4/src/entry.rs#L22

RealKC replied:
hm I guess I could switch to using TryFrom, the main point of this function is to (a) have the comment in one place (b) not repeat the actual expression to go from g_unichar -> char everywhere it's needed

I guess I could mark it as `#[inline]` and `#[track_caller]` though?

bilelmoussaoui replied:
Sounds like a good plan to me, but i would wait for @sdroege's opinion on this one :)

sdroege replied:
Sounds like a plan

sdroege commented on glib/src/unichar.rs:
```diff
@@ -0,0 +1,239 @@
+}
+
+impl Unichar for char {
+    fn unicode_type(self) -> UnicodeType {
```
You probably want to add `#[inline]`  here in in other places for these trivial functions to give some more hints to the compiler.

RealKC commented:
Regarding the CI failures:
- can I tell typos to ignore either a specific file or specific line? I'd rather not ignore the paranthesis typo on the whole project (and where it is made, it is necessary)
- do I need to add doc aliases to the impls for the functions too of the trait too? Seems strange

sdroege commented:
> do I need to add doc aliases to the impls for the functions too of the trait too? Seems strange

You only need to add them on plain functions and trait declarations, not trait implementations.

> can I tell typos to ignore either a specific file or specific line? I'd rather not ignore the paranthesis typo on the whole project (and where it is made, it is necessary)

Something for @bilelmoussaoui I guess :)

RealKC commented:
Changes since last push:
 - rebased on top of master
 - addressed review comments

> You only need to add them on plain functions and trait declarations, not trait implementations.

As far as I can tell, CI is now failing because the aliases don't exist in the trait impl

RealKC commented:
I rewrote the trait so that CI would be happy, seems to have worked

bilelmoussaoui commented on glib/src/unichar.rs:
```diff
@@ -0,0 +1,239 @@
+        };
+
+        // SAFETY: `a` and `b` will always be init after the g_unichar_decompose call returns
+        let (a, b) = unsafe { (a.assume_init(), b.assume_init()) };
```
Even if res is false ?

RealKC replied:
> If ch is not decomposable, a is set to ch and b is set to zero.
> Return value
> TRUE if the character could be decomposed.

Based on these two statements, I'd say yes, but I can move the assignment in the `if res {` branch, it's not used on the else case anyway

sdroege replied:
Yes if `FALSE` is returned than `a` is set to `ch` and `b` is set to `0`. Both values are always initialized.

sdroege commented on .typos.toml:
```diff
@@ -1,5 +1,5 @@
 [files]
-extend-exclude = ["auto", "sys", "gobject-sys", "*.svg"]
+extend-exclude = ["auto", "sys", "gobject-sys", "*.svg", "glib/Gir.toml"]
```
@bilelmoussaoui Is this fine with you?

bilelmoussaoui replied:
I would ask @Hofer-Julian

RealKC replied:
I think the only alternative is ignoring "paranthasis" as a typo in general? Which I don't really like as an idea. I should add a comment explaining why we ignore that file though

Hofer-Julian replied:
Yes, this looks fine to me

sdroege commented on glib/src/unichar.rs:
```diff
@@ -0,0 +1,238 @@
+
+mod sealed {
+    pub trait Sealed: Into<u32> {
+        fn from_unichar(unichar: u32) -> char;
```
This doesn't make much sense. The `Sealed` trait is only for trait privacy reasons, it should not contain any functions or anything. It's just a marker trait.

How about implementing `glib::UnsafeFrom<u32> for char` and requiring that in the `Unichar` trait?

sdroege commented on glib/src/unichar.rs:
```diff
@@ -0,0 +1,238 @@
+/// This trait provides access to Unicode character classification and manipulations functions
+/// provided by GLib that do not exist in the standard library
+#[doc(alias = "g_unichar")]
+pub trait Unichar: sealed::Sealed + Copy {
```
```suggestion
pub trait Unichar: sealed::Sealed + Into<u32> + Copy {
```

sdroege commented on glib/src/unichar.rs:
```diff
@@ -0,0 +1,238 @@
+        fn from_unichar(unichar: u32) -> char {
+            // It's a glib bug if they give us an invalid unicode codepoint from one of their manipulation
+            // functions. It's safest if we just panic in that case
+            char::from_u32(unichar).expect("glib returned an invalid Unicode codepoint")
```
FWIW I would go with `from_u32_unchecked()` here. If you really want the checks here then only enable them in `#[cfg(debug-assertions)]`.

The `expect()` here together with all the panic handling will considerably blow up the code size when working with these functions, and if glib gives us an invalid value then that's a bug in glib. We have to trust it to do the right thing in more critical places already.

sdroege commented on glib/src/unichar.rs:
```diff
@@ -0,0 +1,238 @@
+    #[inline]
+    fn mirror_char(self) -> Option<char> {
+        let mut mirrored = MaybeUninit::uninit();
+        let res = unsafe {
```
The whole function body should really be in an `unsafe` here. The three steps of creating uninitialized memory, calling the function and only assuming the memory is initialized when the function succeeds all belong together and are the same "safety promise". Just try writing a `SAFETY` comment for each of the individual `unsafe` blocks and you'll see what I mean (you'll have to refer to the other `unsafe` blocks!).

sdroege commented on glib/src/unichar.rs:
```diff
@@ -0,0 +1,238 @@
+            )
+        };
+        debug_assert!(decomposition_length <= buffer_len);
+        unsafe { buffer.set_len(decomposition_length) };
```
This `unsafe` block should be together with the other one *and* the assertion. Reasoning further above.

sdroege commented on glib/src/unichar.rs:
```diff
@@ -0,0 +1,238 @@
+
+        if res {
+            // SAFETY: `a` and `b` will always be init after the g_unichar_decompose call returns
+            let (a, b) = unsafe { (a.assume_init(), b.assume_init()) };
```
Same about the `unsafe` block here

sdroege commented on glib/src/unichar.rs:
```diff
@@ -0,0 +1,238 @@
+            unsafe { from_glib(ffi::g_unichar_compose(a.into(), b.into(), out.as_mut_ptr())) };
+
+        if res {
+            Some(Self::from_unichar(unsafe { out.assume_init() }))
```
Same about the `unsafe` block here
